src/Native/Unix/System.Native/pal_io.c

```diff
@@ -1183,30 +1187,80 @@ static int32_t CopyFile_ReadWrite(int inFd, int outFd)
 }
 #endif // !HAVE_FCOPYFILE
 
-int32_t SystemNative_CopyFile(intptr_t sourceFd, intptr_t destinationFd)
+int32_t SystemNative_CopyFile(intptr_t sourceFd, const char* srcPath, const char* destPath, int32_t overwrite)
 {
     int inFd = ToFileDescriptor(sourceFd);
-    int outFd = ToFileDescriptor(destinationFd);
+    int outFd;
+    int ret;
+    struct stat_ sourceStat;
+
+    while ((ret = fstat_(inFd, &sourceStat)) < 0 && errno == EINTR);
+    if (ret != 0)
+    {
+        return -1;
+    }
+
+    struct stat_ destStat;
+    while ((ret = stat_(destPath, &destStat)) < 0 && errno == EINTR);
+    if (ret == 0)
+    {
+        if (sourceStat.st_dev == destStat.st_dev && sourceStat.st_ino == destStat.st_ino)
+        {
+            errno = EBUSY;
+            return -1;
+        }
+
+        if (!overwrite)
+        {
+            errno = EEXIST;
+            return -1;
+        }
+
+#if HAVE_CLONEFILE
+        // For clonefile we need to unlink the destination file first but we need to
+        // check permission first to ensure we don't try to unlink read-only file
+        if (access(destPath, W_OK) == 0)
+        {
+            ret = unlink(destPath);
+            if (ret != 0)
+            {
+                return ret;
+            }
```

Comment:
If the unlink fails, should we instead just let it try to clonefile?  What would the errno be in that case?

Reply:
Actually we should probably bail out early if `access` fails. It prevents unlinking of read-only files. I just forgot to do it.

If the `unlink` fails and we'd ignore the error two things can happen:
- `clonefile` would fail with `EEXIST` error and we would end up in error case anyway. Unlike the error from `unlink` the error code would end up wrong, eg. instead of `EPERM`/`EACCES` you'd get `EEXIST`.
- Someone can unlink the file from different thread/process meanwhile before `clonefile` is called. This is technically a race condition similar to the second `open` call on the destination file failing. Failing here seems no different to how `CopyFile` could fail before.